Fix typeahead behaviour for QueryField

miqh opened the pull request:
I'm new to Grafana (awesome application by the way) and thought I'd take a shot at #13484.

While looking at this bug I actually discovered a few other problems, which turned out to be somewhat related:

* Using a negated label matching operator (e.g. `metric{label!=}`) doesn't cause suggestions for label values to appear. Instead, labels appear as further suggestions, despite already being within a label value context.
* Picking a suggestion which has a non-word leading character when part of the suggestion has already been typed out causes a misaligned replacement. For example, typed out `metric{label="/ro}`, one label value suggestion that appears is `/route`, select `/route`, the replacement results in `metric{label="//route"}`.
* ~~Typing inside a label value pre-enclosed in double quotes will never result in any suggestions.~~

Anyway, I made some changes which appear to resolve all the above, plus the original problem described in the related issue.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=13663) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=13663) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=13663) it.</sub>

miqh committed:
Fix typeahead behaviour for QueryField

These changes were originally intended to address a bug whereby a
suggestion for an already selected label value continues to appear.
However, they also appear to fix several other problems in the area:

- Wrong suggestions when using negated label matching operators
- Misaligned label value suggestion replacements

Related: #13484

miqh committed:
Update PromQueryField tests to address fixed bug

Related: #13484

davkal commented on public/app/features/explore/PromQueryField.tsx:
```diff
@@ -421,7 +421,7 @@ class PromQueryField extends React.PureComponent<PromQueryFieldProps, PromQueryF
     const existingKeys = parsedSelector ? parsedSelector.labelKeys : [];
 
-    if ((text && text.startsWith('=')) || _.includes(wrapperClasses, 'attr-value')) {
+    if ((text && text.match(/^[!=]/)) || _.includes(wrapperClasses, 'attr-value')) {
```
Can we use a similar operator regex as above? Eg. `^!?=~?` ?

miqh replied:
Yep, absolutely—I was being frugal with the match here, seeing as though the label matching operators can only ever start with `!` or `=` (at present anyway).

davkal commented on public/app/features/explore/QueryField.tsx:
```diff
@@ -228,7 +228,13 @@ class QueryField extends React.PureComponent<TypeaheadFieldProps, TypeaheadField
+      // blindly applied, which would undesirably interfere with suggestions
+      const labelValueMatch = prefix.match(/(?:!?=~?"?|")(.*)/);
+      if (labelValueMatch) {
+        prefix = labelValueMatch.pop();
```
Prefer to use group indexes instead of `pop()` as in the rest of the class.

davkal commented on public/app/features/explore/utils/prometheus.ts:
```diff
@@ -66,24 +66,21 @@ export function parseSelector(query: string, cursorOffset = 1): { labelKeys: any
-    const key = match.slice(0, delimiterIndex);
-    const value = match.slice(delimiterIndex + 1, match.length);
-    labels[key] = value;
+  selector.replace(labelRegexp, (_, label, operator, value) => {
```
Neat! Prefer to use `key` instead of `label` here.

davkal reviewed (changes requested):
Really nicely done!

I only request small naming consistency changes.

And you're right, the typeahead code needs lots of work. (We do have more issues 😉 )

miqh commented:
@davkal, thanks!

I've made all three of your requested edits and squashed them into the commit containing the bulk of the actual fixes.

Happy to look at more typeahead-related things if you can think of any to point me toward. :slightly_smiling_face:

davkal commented:
Excellent work! Thank you for your contribution @miqh 

Here are a couple of related issues that we could use some help with: #13604, #13690, #13615
